format/wasm/instructions.go

```diff
@@ -0,0 +1,1539 @@
+	case 1:
+		decodePrefixedInstructionWithoutOperands(d, "i32.trunc_sat_f32_u", v)
+	case 2:
+		decodePrefixedInstructionWithoutOperands(d, "i32.trunc_sat_f64_s", v)
+	case 3:
+		decodePrefixedInstructionWithoutOperands(d, "i32.trunc_sat_f64_u", v)
+	case 4:
+		decodePrefixedInstructionWithoutOperands(d, "i64.trunc_sat_f32_s", v)
+	case 5:
+		decodePrefixedInstructionWithoutOperands(d, "i64.trunc_sat_f32_u", v)
+	case 6:
+		decodePrefixedInstructionWithoutOperands(d, "i64.trunc_sat_f64_s", v)
+	case 7:
+		decodePrefixedInstructionWithoutOperands(d, "i64.trunc_sat_f64_u", v)
+
+	case 8:
+		decodeMemoryInit(d)
+	case 9:
+		decodeDataDrop(d)
+	case 10:
+		decodeMemoryCopy(d)
+	case 11:
+		decodeMemoryFill(d)
+	case 12:
+		decodeTableInit(d)
+	case 13:
+		decodeElemDrop(d)
+	case 14:
+		decodeTableCopy(d)
+	case 15:
+		decodeTableGrow(d)
+	case 16:
+		decodeTableSize(d)
+	case 17:
+		decodeTableFill(d)
+
+	default:
+		d.Fatalf("unknown prefixed instruction: 0xfc %d", v)
+	}
+}
+
+func decodePrefixedInstructionWithoutOperands(d *decode.D, name string, i uint64) {
+	d.FieldUScalarFn(name, readUnsignedLEB128, d.AssertU(i))
+}
+
+func decodeMemoryInit(d *decode.D) {
+	d.FieldUScalarFn("memory.init", readUnsignedLEB128, d.AssertU(8))
+	decodeDataIdx(d, "x")
+	d.FieldU8("reserved", scalar.ActualHex, d.AssertU(0))
+}
+
+func decodeDataDrop(d *decode.D) {
+	d.FieldUScalarFn("data.drop", readUnsignedLEB128, d.AssertU(9))
+	decodeDataIdx(d, "x")
+}
+
+func decodeMemoryCopy(d *decode.D) {
+	d.FieldUScalarFn("memory.copy", readUnsignedLEB128, d.AssertU(10))
+	d.FieldU8("reserved1", scalar.ActualHex, d.AssertU(0))
+	d.FieldU8("reserved2", scalar.ActualHex, d.AssertU(0))
+}
+
+func decodeMemoryFill(d *decode.D) {
+	d.FieldUScalarFn("memory.fill", readUnsignedLEB128, d.AssertU(11))
+	d.FieldU8("reserved", scalar.ActualHex, d.AssertU(0))
+}
+
+func decodeTableInit(d *decode.D) {
+	d.FieldUScalarFn("table.init", readUnsignedLEB128, d.AssertU(12))
+	decodeElemIdx(d, "y")
+	decodeTableIdx(d, "x")
+}
+
+func decodeElemDrop(d *decode.D) {
+	d.FieldUScalarFn("elem.drop", readUnsignedLEB128, d.AssertU(13))
+	decodeElemIdx(d, "x")
+}
+
+func decodeTableCopy(d *decode.D) {
+	d.FieldUScalarFn("table.copy", readUnsignedLEB128, d.AssertU(14))
+	decodeTableIdx(d, "x")
+	decodeTableIdx(d, "y")
+}
+
+func decodeTableGrow(d *decode.D) {
+	d.FieldUScalarFn("table.grow", readUnsignedLEB128, d.AssertU(15))
+	decodeTableIdx(d, "x")
+}
+
+func decodeTableSize(d *decode.D) {
+	d.FieldUScalarFn("table.size", readUnsignedLEB128, d.AssertU(16))
+	decodeTableIdx(d, "x")
+}
+
+func decodeTableFill(d *decode.D) {
+	d.FieldUScalarFn("table.fill", readUnsignedLEB128, d.AssertU(17))
+	decodeTableIdx(d, "x")
+}
+
+func decodeVectorInstruction(d *decode.D) {
+	d.FieldU8("prefix", d.AssertU(0xfd), scalar.ActualHex)
+	s := peekUnsignedLEB128(d)
+	v, ok := s.Actual.(uint64)
+	if !ok {
+		d.Fatalf("expected uint64 but got %t", s.Actual)
+	}
+	switch v {
+	case 0:
```

Comment:
For some formats it has sometimes been helpful to have custom type with names and a decode function and then create a table with decoders. But i like simple switch statements also sometimes.

Some inspiration is https://github.com/wader/fq/blob/master/format/msgpack/msgpack.go and maybe https://github.com/wader/fq/blob/master/format/cbor/cbor.go

Reply:
I like the idea to have such a custom type. I'll consider it.

Reply:
Yeap do so but plain non-fancy-pantsy code is nice sometimes too